### PR TITLE
[BEAM-13872] [Playground] Increase test coverage for the code_processing package

### PR DESCRIPTION
**Pull request by pavel-avilov**
[[BEAM-13872]](https://issues.apache.org/jira/browse/BEAM-13872)
Add unit tests to increase test coverage for the code_processing package

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/validate-runner/README.md)

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by pavel-avilov** [comment]
@ilya-kozyrev @KhaninArtur

**Review by daria-malkova — CHANGES_REQUESTED** [review]

**Review by KhaninArtur — CHANGES_REQUESTED** [review]
The same thing with naming of tests as mentioned in this PR – https://github.com/apache/beam/pull/16826

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 1]
```diff
@@ -99,11 +98,12 @@ func Test_Process(t *testing.T) {
 	if err != nil {
 		panic(err)
 	}
-	sdkEnv, err := environment.ConfigureBeamEnvs(appEnvs.WorkingDir())
+	sdkJavaEnv, err := environment.ConfigureBeamEnvs(appEnvs.WorkingDir())
```
[the remark this answers is not in the document]
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go:106]: Done.

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 2]
```diff
@@ -948,11 +967,17 @@ func Test_runStep(t *testing.T) {
...
```
[the remark this answers is not in the document]
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go]: Done

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 3]
```diff
@@ -1004,3 +1180,354 @@ func syncMapLen(syncMap *sync.Map) int {
...
+		cacheService         cache.Cache
+		graphFilePath        string
+		pipelineId           uuid.UUID
+	}
+	tests := []struct {
+		name string
+		args args
+	}{
+		{
+			name: "All success",
```
Comment (KhaninArtur) [inline playground/backend/internal/code_processing/code_processing_test.go:1515]: Do we have a negative test case here?

**Review by KhaninArtur — COMMENTED** [review]
Overall LGTM, just one small question

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 3 above)
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go:1515]: No, we don't have

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 4]
```diff
@@ -61,16 +68,26 @@ func TestMain(m *testing.M) {
 
 func setup() {
 	// create configs for java
-	err := os.MkdirAll("configs", fs.ModePerm)
+	err := os.MkdirAll(configFolder, fs.ModePerm)
 	if err != nil {
 		panic(err)
 	}
-	filePath := filepath.Join("configs", pb.Sdk_SDK_JAVA.String()+".json")
+	filePath := filepath.Join(configFolder, pb.Sdk_SDK_JAVA.String()+".json")
```
Comment (daria-malkova) [inline playground/backend/internal/code_processing/code_processing_test.go:75]: ".json" -> constant

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 5]
```diff
@@ -212,26 +234,45 @@ func Test_Process(t *testing.T) {
...
+				pipelineId:      uuid.New(),
+				pipelineOptions: "",
+			},
+		},
+		{
+			// Test case with calling processCode method with incorrect go code.
+			// As a result status into cache should be set as Status_STATUS_PREPARATION_ERROR.
+			name:                  "Prepare step failed",
+			createExecFile:        true,
+			code:                  "package main\nimport \"fmt\"\nfunc main() {\n    fmt.Println(\"hello world\").\n}\n",
```
Comment (daria-malkova) [inline playground/backend/internal/code_processing/code_processing_test.go:266]: code -> constant

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 6]
```diff
@@ -908,51 +1009,84 @@ func Test_compileStep(t *testing.T) {
...
-		want *executors.Executor
-		code string
+		name           string
+		args           args
+		code           string
+		expectedStatus pb.Status
 	}{
 		{
-			name: "Test compilation step working without an error",
+			name: "Test compilation step working on java sdk",
```
Comment (daria-malkova) [inline playground/backend/internal/code_processing/code_processing_test.go:1018]: working -> finishes successfully

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 7]
```diff
@@ -908,51 +1009,84 @@ func Test_compileStep(t *testing.T) {
...
+				sdkEnv:               sdkJavaEnv,
+				isUnitTest:           false,
+				pipelineLifeCycleCtx: context.Background(),
+				cancelChannel:        make(chan bool, 1),
+			},
+			code:           helloWordJava,
+			expectedStatus: pb.Status_STATUS_EXECUTING,
+		},
+		{
+			name: "Test compilation step working on python sdk",
```
Comment (daria-malkova) [inline playground/backend/internal/code_processing/code_processing_test.go:1032]: ditto

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 8]
```diff
@@ -962,36 +1096,78 @@ func Test_runStep(t *testing.T) {
...
-		args args
-		code string
+		name           string
+		args           args
+		code           string
+		expectedStatus pb.Status
 	}{
 		{
-			name: "Test run step working without an error",
+			name: "Test run step working on python sdk",
```
Comment (daria-malkova) [inline playground/backend/internal/code_processing/code_processing_test.go:1108]: ditto

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 9]
```diff
@@ -962,36 +1096,78 @@ func Test_runStep(t *testing.T) {
...
+				pipelineOptions:      "",
+				pipelineLifeCycleCtx: context.Background(),
+				cancelChannel:        make(chan bool, 1),
+				createExecFile:       true,
+			},
+			code:           helloWordPython,
+			expectedStatus: pb.Status_STATUS_RUN_ERROR,
+		},
+		{
+			name: "Test run step working on go sdk",
```
Comment (daria-malkova) [inline playground/backend/internal/code_processing/code_processing_test.go:1124]: ditto

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 10]
```diff
@@ -962,36 +1096,78 @@ func Test_runStep(t *testing.T) {
...
+				pipelineOptions:      "",
+				pipelineLifeCycleCtx: context.Background(),
+				cancelChannel:        make(chan bool, 1),
+				createExecFile:       true,
+			},
+			code:           helloWordGo,
+			expectedStatus: pb.Status_STATUS_RUN_ERROR,
+		},
+		{
+			name: "Test run step without preparing files with code",
```
Comment (daria-malkova) [inline playground/backend/internal/code_processing/code_processing_test.go:1140]: success or fail?

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 3 above)
Comment (daria-malkova) [inline playground/backend/internal/code_processing/code_processing_test.go:1515]: Let's have more descriptive name

**Review by daria-malkova — CHANGES_REQUESTED** [review]

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 5 above)
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go:266]: Moved to a variable, not to constant

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 4 above)
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go:75]: Done

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 6 above)
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go:1018]: Done

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 7 above)
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go:1032]: Done

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 8 above)
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go:1108]: Added commentaries

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 9 above)
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go:1124]: Added commentaries

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 10 above)
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go:1140]: Fail, expected to get `Status_STATUS_RUN_ERROR` pipeline status

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 3 above)
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go]: Done

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 11]
```diff
@@ -61,16 +69,26 @@ func TestMain(m *testing.M) {
 
 func setup() {
 	// create configs for java
-	err := os.MkdirAll("configs", fs.ModePerm)
+	err := os.MkdirAll(configFolder, fs.ModePerm)
 	if err != nil {
 		panic(err)
 	}
-	filePath := filepath.Join("configs", pb.Sdk_SDK_JAVA.String()+".json")
+	filePath := filepath.Join(configFolder, pb.Sdk_SDK_JAVA.String()+jsonExtension)
```
Comment (AydarZaynutdinov) [inline playground/backend/internal/code_processing/code_processing_test.go:76]:
```suggestion
	filePath := filepath.Join(configFolder, fmt.Sprintf("%s%s", pb.Sdk_SDK_JAVA.String(), jsonExtension))
```

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 12]
```diff
@@ -587,12 +630,25 @@ func Test_getRunOrTestCmd(t *testing.T) {
...
+	return sdkEnv, nil
+}
+
+func setupSDK(sdk pb.Sdk) {
 	err := os.MkdirAll(configFolder, fs.ModePerm)
 	if err != nil {
 		panic(err)
 	}
-	filePath := filepath.Join(configFolder, sdk.String()+".json")
+	filePath := filepath.Join(configFolder, sdk.String()+jsonExtension)
```
Comment (AydarZaynutdinov) [inline playground/backend/internal/code_processing/code_processing_test.go:651]:
```suggestion
	filePath := filepath.Join(configFolder, fmt.Sprintf("%s%s", pb.Sdk_SDK_JAVA.String(), jsonExtension))
```

**Review by AydarZaynutdinov — APPROVED** [review]
Overall, LGTM. Just small suggestions.

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 11 above)
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go]: Done

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 12 above)
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go]: Done

File: playground/backend/internal/code_processing/code_processing_test.go [hunk 13]
```diff
@@ -1004,3 +1190,354 @@ func syncMapLen(syncMap *sync.Map) int {
...
+	}{
+		{
+			name: "Successfully saving the prepared graph to the cache",
+			args: args{
+				pipelineLifeCycleCtx: pipelineLifeCycleCtx,
+				backgroundCtx:        context.Background(),
+				cacheService:         cacheService,
+				graphFilePath:        graphFilePath,
+				pipelineId:           uuid.New(),
+			},
```
Comment (KhaninArtur) [inline playground/backend/internal/code_processing/code_processing_test.go:1532]: What happens when we don't have the file with a graph? Does any error raise?

**Review by KhaninArtur — CHANGES_REQUESTED** [review]

File: playground/backend/internal/code_processing/code_processing_test.go (hunk 13 above)
Comment (pavel-avilov) [reply playground/backend/internal/code_processing/code_processing_test.go:1532]: There are no errors raised. If we don't have a graph file, we'll just skip saving it to the cache.

**Review by KhaninArtur — APPROVED** [review]
LGTM

**Comment by pavel-avilov** [comment]
R: @pabloem

**Comment by pabloem** [comment]
LGTM

**Pull request merged**